contracts/pre-propose/cwd-pre-propose-approver/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+    }
+
+    // On completion send rejection or approval message
+    let msg = match new_status {
+        Status::Closed => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Reject { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        Status::Executed => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Approve { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        Status::Passed => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Reject { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        Status::Rejected => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Approve { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        // TODO Maybe don't default reject?
+        _ => None,
+    };
+
+    // TODO two match expressions annoys me
+    match msg {
+        Some(msg) => Ok(Response::default().add_message(msg)),
+        // TODO Not sure what we want to do here...
+        None => Err(PreProposeError::Unauthorized {}),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::QueryExtension { msg } => match msg {
+            QueryExt::PreProposeApprovalContract {} => {
+                to_binary(&PRE_PROPOSE_APPROVAL_CONTRACT.load(deps.storage)?)
+            }
+        },
+        _ => PrePropose::default().query(deps, env, msg),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, PreProposeError> {
+    match msg.id {
+        APPROVER_PROPOSE_REPLY_ID => {
+            match msg.result {
+                SubMsgResult::Ok(result) => {
+                    let pre_propose_id = CURRENT_PRE_PROPOSE_ID.load(deps.storage)?;
+
+                    // Parse SubMessageResponse to get proposal id
+                    let event = &result.events[result.events.len() - 1];
+                    let proposal_id = event.attributes[event.attributes.len() - 1]
+                        .value
+                        .parse::<u64>()?;
+
+                    // Save mapping of proposal id to pre propose id
+                    PROPOSAL_IDS.save(deps.storage, proposal_id, &pre_propose_id)?;
```

Comment:
i think a less error prone way to get this mapping would be to handle the `ExecuteMsg::ProposalCompletedHook`. that way you don't have to dispatch this as a submessage and read the responses. that hook has information about the proposal id.

Reply:
Having trouble getting that hook to fire on proposal passed or rejected... need to investigate...

Reply:
I think we need this mapping of the proposal id in the approver DAO to the pre-proposal id in the DAO seeking approval... can't think of another way to get it as `ExecuteMsg::ProposalCompletedHook` is only executed when the proposal is completed.